### PR TITLE
Add ability to customize template keys

### PR DESCRIPTION
**Pull request by redzic**
This PR makes the `pos` and `len` fields of `ProgressState` to have `pub` visibility. I changed the `per_sec` method of `ProgressState` to return an `f64`, and `{per_sec}` in the template now rounds to an integer via the `format!` macro (rather than truncating by default), and is explicitly truncated in other places of the code. With this patch, you can now use `.key(...)` method from `ProgressStyle` to add a custom formatting function and template key. This PR also adds another example to show the usage of this feature.

I also changed the `position` method to just be 2 separate methods for returning the position and length, because the usage beforehand where `position()` was called before matching the template keys and then you would just reuse those for `{pos}` and `{len}` now seems like an anti-pattern with this hash map based approach.  

Also I had to remove `#[derive(Debug)]` for `ProgressStyle` because I was getting errors of "`Debug` impl is not general enough" or something like that, and it didn't seem to be necessary anyway.

### CODE REVIEW

File: examples/custom-template.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+use std::thread;
```
Comment (djc) [inline examples/custom-template.rs:1]: Please use the new feature in an existing example, and please do it in a separate commit.

File: src/progress_bar.rs [hunk 2]
```diff
@@ -187,7 +187,7 @@ impl ProgressBar {
     pub fn set_draw_rate(&self, n: u64) {
         let mut state = self.state.lock().unwrap();
         state.draw_rate = n;
-        state.draw_next = state.pos.saturating_add(state.per_sec() / n);
+        state.draw_next = state.pos.saturating_add(state.per_sec() as u64 / n);
```
Comment (djc) [inline src/progress_bar.rs:190]: I think we would get better data if the conversion to `u64` wraps the division, right?

File: src/state.rs [hunk 3]
```diff
@@ -82,9 +82,14 @@ impl ProgressState {
         pct.max(0.0).min(1.0)
     }
 
-    /// Returns the position of the status bar as `(pos, len)` tuple.
-    pub fn position(&self) -> (u64, u64) {
-        (self.pos, self.len)
+    /// Returns the position of the status bar.
```
Comment (djc) [inline src/state.rs:85]: If we're going to make the fields public we should just get rid of getter(s).

File: src/state.rs [hunk 4]
```diff
@@ -120,12 +125,12 @@ impl ProgressState {
     }
 
     /// The number of steps per second
-    pub fn per_sec(&self) -> u64 {
-        let avg_time = self.est.seconds_per_step();
-        if avg_time == 0.0 {
-            0
+    pub fn per_sec(&self) -> f64 {
+        let per_sec = 1.0 / self.est.seconds_per_step();
```
Comment (djc) [inline src/state.rs:129]: Why did you change the structure like this? I think I prefer comparing to zero rather than `is_nan()`.

File: src/state.rs [hunk 5]
```diff
@@ -157,7 +162,7 @@ impl ProgressState {
         }
         if new_pos >= self.draw_next {
             self.draw_next = new_pos.saturating_add(if self.draw_rate != 0 {
-                self.per_sec() / self.draw_rate
+                self.per_sec() as u64 / self.draw_rate
```
Comment (djc) [inline src/state.rs:165]: Cast after division?

File: src/style.rs [hunk 6]
```diff
@@ -4,19 +4,81 @@ use crate::format::{BinaryBytes, DecimalBytes, FormattedDuration, HumanBytes, Hu
 use crate::state::ProgressState;
 use crate::utils::{expand_template, pad_str};
 use std::borrow::Cow;
+use std::collections::HashMap;
 
 #[cfg(feature = "improved_unicode")]
 use unicode_segmentation::UnicodeSegmentation;
 
+pub type Format = fn(&ProgressState) -> String;
```
Comment (djc) [inline src/style.rs:12]: I don't think the `FormatMap` wrapper is useful here, since it's private to the crate anyway. Let's just rename `FormatMapInner` to `FormatMap`.

File: src/style.rs [hunk 7]
```diff
@@ -129,6 +193,12 @@ impl ProgressStyle {
         self
     }
 
+    /// Adds a custom key parser that references a `&ProgressState` to the template.
+    pub fn key(mut self, key: &'static str, f: Format) -> ProgressStyle {
```
Comment (djc) [inline src/style.rs:197]: Let's call this `with_key()` (and please remove the period from the docstring).

File: src/style.rs [hunk 8]
```diff
@@ -4,19 +4,81 @@ use crate::format::{BinaryBytes, DecimalBytes, FormattedDuration, HumanBytes, Hu
...
 #[cfg(feature = "improved_unicode")]
 use unicode_segmentation::UnicodeSegmentation;
 
+pub type Format = fn(&ProgressState) -> String;
+type FormatMapInner = HashMap<&'static str, Format>;
+
+#[derive(Clone)]
+pub(crate) struct FormatMap(FormatMapInner);
+
+impl Default for FormatMap {
```
Comment (djc) [inline src/style.rs:18]: I think we should leave the built-in formatters outside of the format map, the map should be used only to augment/replace the default set of keys. In `expand_template()`, let's cache whether the map is empty and only do lookups if it isn't.

**Review by djc — COMMENTED** [review]
Looking pretty good, I have a bunch of smallish suggestions.

File: src/state.rs (hunk 4 above)
Comment (redzic) [reply src/state.rs:129]: Two reasons: even when comparing the denominator to zero, I sometimes get `NaN` instead of 0 as the result anyway. Probably due to using `==` on floating point values. So, semantically, I think that comparing to `NaN` is more accurate.

Other reason is that the codegen seems a bit better when comparing to `NaN`. The compiler only generates branchless code when comparing to NaN, and there are fewer instructions overall. https://godbolt.org/z/f4fhPs7ec

Comment (djc) [reply src/state.rs:129]: Sounds good!

File: examples/custom-template.rs (hunk 1 above)
Comment (redzic) [reply examples/custom-template.rs]: Thanks, deleted the example for now. Will modify an existing example in another commit later.

File: src/progress_bar.rs (hunk 2 above)
Comment (redzic) [reply src/progress_bar.rs]: Fixed. I cast both values to `f64` now and then cast to `u64` at the end.

File: src/state.rs (hunk 3 above)
Comment (redzic) [reply src/state.rs]: Removed these methods.

File: src/state.rs (hunk 5 above)
Comment (redzic) [reply src/state.rs]: Thanks, fixed.

File: src/style.rs (hunk 7 above)
Comment (redzic) [reply src/style.rs]: Done.

File: src/style.rs (hunk 6 above)
Comment (redzic) [reply src/style.rs:12]: Removed `FormatMapInner`.

File: src/style.rs (hunk 8 above)
Comment (redzic) [reply src/style.rs:17]: Alright, that makes sense. However, I'm not really sure how you would do this in `expand_template` as it is right now. Should `expand_template` also take a reference to the hash map of custom keys as well? Also, how would it know if the key was not found without changing the `format_state` function? Currently, it just falls back to returning an empty string and just using that in the template. Perhaps this could be changed to return an `Option<String>`?

To me, it seems easier to just do all of that inside of the closure in `format_state`. It doesn't really seem necessary to cache whether the map is empty, as it seems very cheap to check for that by just calling `.is_empty()`: https://godbolt.org/z/M5fEPjWE4

Comment (djc) [reply src/style.rs:17]: Ah, sorry, I guess I mean sort of like it is now, in the closure passed to `expand_template()`.

Comment (redzic) [reply src/style.rs:17]: Ok, will do. One implication of that approach is that the default keys cannot be overridden, because they are always matched first. Perhaps this should be mentioned in the documentation?

Comment (redzic) [reply src/style.rs:17]: The fact that default template keys are not overridable is now mentioned in the docstring of `with_key`.

**Comment by redzic** [comment]
@djc should be ready to review

File: src/progress_bar.rs [hunk 9]
```diff
@@ -478,7 +480,7 @@ impl ProgressBar {
 
     /// Returns the current rate of progress
     pub fn per_sec(&self) -> u64 {
-        self.state.lock().unwrap().per_sec()
+        self.state.lock().unwrap().per_sec() as u64
```
Comment (djc) [inline src/progress_bar.rs:483]: How do you feel about just making this also return `f64`?

File: src/style.rs [hunk 10]
```diff
@@ -261,13 +283,24 @@ impl ProgressStyle {
...
-                "bytes_per_sec" => format!("{}/s", HumanBytes(state.per_sec())),
-                "binary_bytes_per_sec" => format!("{}/s", BinaryBytes(state.per_sec())),
+                "bytes_per_sec" => format!("{}/s", HumanBytes(state.per_sec() as u64)),
+                "binary_bytes_per_sec" => format!("{}/s", BinaryBytes(state.per_sec() as u64)),
                 "eta_precise" => format!("{}", FormattedDuration(state.eta())),
                 "eta" => format!("{:#}", HumanDuration(state.eta())),
                 "duration_precise" => format!("{}", FormattedDuration(state.duration())),
                 "duration" => format!("{:#}", HumanDuration(state.duration())),
-                _ => "".into(),
+                key => {
```
Comment (djc) [inline src/style.rs:292]: Put together like this, the separate `format_map.0.is_empty()` case looks silly. Let's assume that `.get()` will be reasonably quick in the case of a zero-length map.

I propose that we do `if let Some(_) = self.format.map.0.get(var)` before the `match var.key`, so that we can override the default keys.

Comment (redzic) [reply src/style.rs:292]: Seems reasonable. Hash table lookups, especially on small strings, should be quite fast.

File: src/progress_bar.rs (hunk 9 above)
Comment (redzic) [reply src/progress_bar.rs:483]: `f64` is probably more consistent, so I think it'd be better.

File: src/style.rs (hunk 8 above)
Comment (redzic) [reply src/style.rs:17]: This change has been reverted, as the default keys are now overridable.
